### ISSUE TITLE
Possible problem with foreach loop?

### ISSUE THREAD

**Issue by MafaldaSFerreira** [melisaolave/SNPs2CF#1]
Hello,

I have been excited to try your SNPs2CF script after I saw your talk at Evolution! Thanks for sharing it with the community.

I was attempting to run it on a 45779 SNP dataset with 16 species and 34 individuals (so, multiple individuals per species).

First, I tried to run SNPs2CF with between.sp.only=TRUE and I found no problems. 

```
output <- SNPs2CF(seqMatrix="snp_alignment_34indvs.phy", between.sp.only = TRUE,bootstrap=FALSE,outputName="SNPs2CFsnp_alignment_34indvs.csv",ImapName="map.txt",cores=50);
```

However, when I tried to set between.sp.only to FALSE, taking advantage of all individual quartet information, the script halts at nearly the end of analyzing all the possible quartets (I say this because I was running within a screen session, and I can't scroll back enough to see if it reached quartet 46376th, but the last quartet was 46337). 

This is the code:
```
output <- SNPs2CF(seqMatrix="snp_alignment_34indvs.phy", between.sp.only = FALSE,bootstrap=FALSE, outputName="SNPs2CFsnp_alignment_34indvs_allQua.csv",ImapName="map.txt",cores=50);
```

What I see is this error:
```
...
A total of 12 individual quartets will be sampled within each of the 46376 species quartets
Progress:
         1 / 12 of 46337 / 46376
         2 / 12 of 46337 / 46376
         3 / 12 of 46337 / 46376
         4 / 12 of 46337 / 46376
         5 / 12 of 46337 / 46376
         6 / 12 of 46337 / 46376
         7 / 12 of 46337 / 46376
         8 / 12 of 46337 / 46376
         9 / 12 of 46337 / 46376
         10 / 12 of 46337 / 46376
         11 / 12 of 46337 / 46376
         12 / 12 of 46337 / 46376
Error in { :
  task 5457 failed - "no loop for break/next, jumping to top level"
Calls: SNPs2CF -> %dopar% -> <Anonymous>
Execution halted
```

I tried to search about this and it seems it could have something to do with the parallelization with foreach? https://stackoverflow.com/questions/35273971/dopar-error-after-several-iterations
Do you have any ideia what could be happening?

Thank you,
Mafalda

**Comment by melisaolave** [comment]
Hello Mafalda, 
I've been woking on this but yet couldn't find where the problem is. 
However, this is something that could help you: I just added some changes to the SNPs2CF function and it now allows to save progress and continue from an specified quartet number. See the new tutorial v1.2 "Saving progress, continuing calculations and combining temporal CF tables" for an explanation.

If you want to save some time (while I find out what's wrong), you could try to run the function again leaving save.progress=TRUE and when the error comes out, rerun the function setting starting.sp.quartet= 46378 (that's one after the error comes out). Then you will need to use the function combine.CF.table() to append all the independent outputs (also described in the new tutorial)

You will have then a table with almost all quartets, suitable for PhyloNetworks.

PS: I have the impression that 46K quartets is too much for PhyloNetworks. You might be able to get a network estimation within a couple of weeks of analyses, but boostrap calculation will (most likely) take too long, and become unrealistic. You can try, though.

**Comment by MafaldaSFerreira** [comment]
Hello Melisa,

 

I’m sorry for my late response. I have been doing lab work and did not have time to look at this again until now! Thank you for your help and response.
I will maybe try to use your solution. However, I agree with you that PhyloNetworks with my raw dataset might not be the way to go. As I told you, I generated a CF table with species only SNPs and the snaq run hasn’t finished yet… I think it might be impractical. I might try to run a very small subset of species for which I have species introgression hypothesis from other analysis. Thank you for your advice! What do you think is a practicable number of quartets for PhyloNetworks? 

Thank you very much,
Mafalda

 

De: Melisa Olave <notifications@github.com>
Responder A: melisaolave/SNPs2CF <reply@reply.github.com>
Data: segunda-feira, 8 de outubro de 2018, 12:01
Para: melisaolave/SNPs2CF <SNPs2CF@noreply.github.com>
Cc: Mafalda Ferreira <amafaldasferreira@gmail.com>, Author <author@noreply.github.com>
Assunto: Re: [melisaolave/SNPs2CF] Possible problem with foreach loop? (#1)

 

Hello Mafalda,
I've been woking on this but yet couldn't find where the problem is.
However, this is something that could help you: I just added some changes to the SNPs2CF function and it now allows to save progress and continue from an specified quartet number. See the new tutorial v1.2 "Saving progress, continuing calculations and combining temporal CF tables" for an explanation.

If you want to save some time (while I find out what's wrong), you could try to run the function again leaving save.progress=TRUE and when the error comes out, rerun the function setting starting.sp.quartet= 46378 (that's one after the error comes out). Then you will need to use the function combine.CF.table() to append all the independent outputs (also described in the new tutorial)

You will have then a table with almost all quartets, suitable for PhyloNetworks.

PS: I have the impression that 46K quartets is too much for PhyloNetworks. You might be able to get a network estimation within a couple of weeks of analyses, but boostrap calculation will (most likely) take too long, and become unrealistic. You can try, though.

**Comment by melisaolave** [comment]
Hi Mafalda, 

About time vs. number of quartets, it of course depends in your dataset, number of species, how many hybridizing edges you are testing for, and the computational resources available (among others). Running smaller clades in different analyses will help a lot (as you are suggesting). If you can run 2 networks with more or less 8 species, it should work perfectly. 

As an alternative, it is also possible to randomly subsample the individuals using n.quartets element in SNPs2CF function --- check out n.quartets information at the documentation. I would say not to exceed 25,000 quartets. In your case, you have 16 species, then setting n.quartets=14 should give about 25K quartets (if I'm not wrong). But this is still a very large matrix and will take time. You could try it, and if still takes long, decrease the number of n.quartets.

Or you could go with something safer and run n.quartets=5 (those should be about 9,100 quartets, I think). 

And have in mind that the main limitation are the bootstrap calculations, that have to be run after the obtaining the best network. Currently (and to my knowledge), PhyloNetworks only allows to perform independent network estimation in parallel within each bootstrap replicate (e.g. if you are running runs=10 the max number of cores you can take advantage of, it is just 10).  If you really want to use a large number of quartets, the bootstrap calculations can be handled more or less well if you have lots of cores available and some basic coding skills. E.g. running each bootstrap replicate in a separate analyses (then 10 cores are used per each replicate). Later, combining them into a single file is needed. 

Melisa

**Comment by MafaldaSFerreira** [comment]
Hi Melisa,

 

These are good suggestions! Thank you very much for taking the time.

I think I am going to experiment with down sampling my dataset for now, and focusing on specific hypothesis with less taxa. I do have multiple cores available, but even so, a single run of snaq is taking too long with my full dataset. I’m going to wait for it to end, but I believe it has been running for 1-2weeks now. 

 

Thank you for your time Melisa!
Best,
Mafalda

 

De: Melisa Olave <notifications@github.com>
Responder A: melisaolave/SNPs2CF <reply@reply.github.com>
Data: terça-feira, 16 de outubro de 2018, 13:06
Para: melisaolave/SNPs2CF <SNPs2CF@noreply.github.com>
Cc: Mafalda Ferreira <amafaldasferreira@gmail.com>, Author <author@noreply.github.com>
Assunto: Re: [melisaolave/SNPs2CF] Possible problem with foreach loop? (#1)

 

Hi Mafalda,

About time vs. number of quartets, it of course depends in your dataset, number of species, how many hybridizing edges you are testing for, and the computational resources available (among others). Running smaller clades in different analyses will help a lot (as you are suggesting). If you can run 2 networks with more or less 8 species, it should work perfectly.

As an alternative, it is also possible to randomly subsample the individuals using n.quartets element in SNPs2CF function --- check out n.quartets information at the documentation. I would say not to exceed 25,000 quartets. In your case, you have 16 species, then setting n.quartets=14 should give about 25K quartets (if I'm not wrong). But this is still a very large matrix and will take time. You could try it, and if still takes long, decrease the number of n.quartets.

Or you could go with something safer and run n.quartets=5 (those should be about 9,100 quartets, I think).

And have in mind that the main limitation are the bootstrap calculations, that have to be run after the obtaining the best network. Currently (and to my knowledge), PhyloNetworks only allows to perform independent network estimation in parallel within each bootstrap replicate (e.g. if you are running runs=10 the max number of cores you can take advantage of, it is just 10). If you really want to use a large number of quartets, the bootstrap calculations can be handled more or less well if you have lots of cores available and some basic coding skills. E.g. running each bootstrap replicate in a separate analyses (then 10 cores are used per each replicate). Later, combining them into a single file is needed.

Melisa

**Comment by jferrerobiol** [comment]
Hi Melisa,

Thank you very much for having written these useful functions. I am trying to use them with my data and I have had similar issues to Mafalda. Basically, the function SNPs2CF looks like it runs fine but when it gets to the end it raises an error:

error "missing values and NaN's not allowed if 'na.rm' is FALSE"

I have also noticed that there are some quartets that don't appear in the temp list and it looks like it skipped them but when I try to restart the run with starting.sp.quartet= it prints the process in the screen as if it processed them properly but then they still don't appear in the temp directory.

Do you have any hints of why is this happening?

Here it is my code:

`Puffinus_ddRAD <- SNPs2CF(seqMatrix = "Shearwaters_clust89_percent65_1snpxlocus.min4.phy", ImapName = "Puffinus_ddRAD.imap", outputName = "Puffinus_ddRAD_SNPs2CF.csv", n.quartets = 3, save.progress = T, cores = 4)`

I have a dataset with 29 individuals of 15 species with 1-3 individuals per species and I would like to be able to use the Multiple Alleles version of PhyloNetworks, so here I am generating the individual CF table to input it to Phylonetworks to compute the species CF table. The number of quartets to process is 20,475.

Thank you very much for your help,

Joan

**Comment by melisaolave** [comment]
Hello Joan,

I believe that this issue is related to not having any SNP informing CF for one or more species quartets. This means that for some quartets there are not variants in your dataset. Then, there is not an individual quartet to save in the temporal folder.
I should fix this so the function returns a more informative error. I promise to do it when I find time.

BUT even after adding a warning to the function, you will still have this issue in your dataset. In your case, you have multiple individuals per species, and you are setting n.quartets = 3, then there should be 3 individual quartet per species quartet. If you at least have one individual quartet per species quartet that is informative, then my guess is that it should be fine for PhyloNetworks.
Instead, if there are not representatives for a species quartet (i.e. all 3 individual quartets are have not informative SNPs), then I'm not sure if it would affect network estimation or not. If this is your case, you could ask PhyloNetworks developers whether they think it might impact your results, but I don't think that case was explored so far. 

Melisa

**Comment by melisaolave** [comment]
Alright,

since version 1.2, the new save.progress and starting.sp.quartet objects can be used to re-start analysis from a given point, so potential RAM issues can be more or less addressed (i.e. if it crashes, just re-start the analysis from the last analyzed quartet, then combine tables using combine.CF.table function). See tutorial.

Also, the error "missing values and NaN's not allowed if 'na.rm' is FALSE", is now solved in the version 1.3. This was due to lack of informative SNPs during bootstrap calculations. The function will also report a warning about which quartet is failing.
